src/decomp/decompstreambuffer.cpp

```diff
@@ -0,0 +1,209 @@
+    if (mode & std::ios::out)
+        *fmodeptr++ = 'w';
+    *fmodeptr++ = 'b';
+    *fmodeptr = '\0';
+
+    //printf("filemode %s\n", fmode);
+
+    char data_name[1024];
+    strcpy(data_name, name);
+    strcat(data_name, ".data");
+    data = lzmaopen(data_name, fmode);
+    char meta_name[1024];
+    strcpy(meta_name, name);
+    strcat(meta_name, ".meta");
+    meta = lzmaopen(meta_name, fmode);
+
+    if (meta == 0)
+        return 0;
+
+    opened = true;
+
+    printf("dcmpin triggered\n");
+
+    meta_processed = 0;
+    data_processed = 0;
+
+    // TODO: read compressedsections from meta
+    printf("dcmpin stage PARSE COMPRESSEDSECTIONS\n");
+
+    int num = 1;
+    int current = 0;
+    int status = BUFFER_SHORT;
+    uint64_t value = 0;
+    int8_t size = 0;
+
+    while (num != 0 && status != FINISHED) {
+        if (status == NOT_OK) {
+            printf("AAAAAAAAAAAAAAAAAAAAAH\n");
+        }
+
+        // read new long long into value
+        num = lzmaread(&value, sizeof(uint64_t), 1, meta);
+        //printf("read %d into buffer\n", num);
+        if (num < 0) {
+            status = NOT_OK;
+            continue;
+        }
+        status = OK;
+
+        // check for terminating sequence
+        if (value == UINT64_MAX) {
+            status = FINISHED;
+            continue;
+        }
+
+        // add compressed section
+
+        // this still needs to be checked for validity, as it stands ofc
+        // parses 8 bytes to a long, checks against -1 BUT parses this
+        // to a uint later, which can lead to overflow errors when left
+        // as a checked expression (and just wraps around if unchecked)
+        // so ultimately, I would need to know if I can just use the first
+        // 4 bytes or if I have to do any sort of specific parsing
+
+        // EDIT: seems ok
+
+        num = lzmaread(&size, sizeof(int8_t), 1, meta);
+        //printf("read %d into buffer\n", num);
+        if (num < 0) {
+            status = NOT_OK;
+            continue;
+        }
+
+        //printf("value %u size %d\n", value, size);
+
+        sections.push_back(CompressedSection((uint32_t) value, 0, size));
+    }
+
+    printf("dcmpin stage PARSE COMPRESSEDSECTIONS finished\n");
+    printf("for a total of %d compressed sections\n", sections.size());
+
+    // start actual decompression
+    if (data == 0) {
+        return 0;
+    }
+
+    decompression = make_unique<DecompressionParser>(data, meta);
+
+    return this;
+}
+
+DecompStreamBuffer* DecompStreamBuffer::close()
+{
+    if (is_open() && (mode & std::ios::out)) {
+        last = true;
+        sync();
+        opened = false;
+        if (lzmaclose(meta) == 0 && lzmaclose(data))
+            return this;
+    }
+    return 0;
+}
+
+int DecompStreamBuffer::underflow()
+{
+    printf("\nunderflow ocurred\n");
+
+    if (!(mode & std::ios::in) || !opened)
+        return EOF;
+
+    int num = decompression->fill_buffer(buffer, bufferSize, sections);
```

Comment:
how about giving "sections" vector to decompression via constructor, instead of having to hold the reference in decompstreambuffer?

Reply:
good idea, will do